File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.py
import sys
INF = sys.maxsize


def solution(alp, cop, problems):
    target_al, target_co = 0, 0
    for problem in problems:
        target_al = max(target_al, problem[0])
        target_co = max(target_co, problem[1])

    dp = [[INF for _ in range(target_co + 1)] for __ in range(target_al + 1)]
    
    alp = min(alp, target_al)
    cop = min(cop, target_co)
    dp[alp][cop] = 0

    for i in range(alp, target_al + 1):
        for j in range(cop, target_co + 1):
            if i < target_al:
                dp[i + 1][j] = min(dp[i + 1][j], dp[i][j] + 1)
            if j < target_co:
                dp[i][j + 1] = min(dp[i][j + 1], dp[i][j] + 1)
            for alp_req, cop_req, alp_rwd, cop_rwd, cost in problems:
                if i >= alp_req and j >= cop_req:
                    tmp_i = min(i + alp_rwd, target_al)
                    tmp_j = min(target_co, j + cop_rwd)
                    dp[tmp_i][tmp_j] = min(dp[tmp_i][tmp_j], dp[i][j] + cost)

    return dp[target_al][target_co]

File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
class Solution {

	int targetAl = 0;
	int targetCo = 0;
    public int solution(int alp, int cop, int[][] problems) {
        int answer = 0;

		for (int[] problem: problems){
			targetAl = Math.max(targetAl, problem[0]);
			targetCo = Math.max(targetCo, problem[1]);
		}

		int[][] dp = new int[targetAl + 1][targetCo + 1];
        
        for (int i = 0; i < targetAl + 1; i++) {
			for (int j = 0; j < targetCo + 1; j++) {
				dp[i][j] = 10000000;
			}
		}

		alp = Math.min(alp, targetAl);
		cop = Math.min(cop, targetCo);

		dp[alp][cop] = 0;

		for (int i = alp; i < targetAl + 1; i++) {
			for (int j = cop; j < targetCo + 1; j++) {

				if (i < targetAl) dp[i + 1][j] = Math.min(dp[i + 1][j], dp[i][j] + 1);
				if (j < targetCo) dp[i][j + 1] = Math.min(dp[i][j + 1], dp[i][j] + 1);

				for (int[] problem: problems) {
					if (i >= problem[0] && j >= problem[1]){
						int tmpI = Math.min(targetAl, i + problem[2]);
						int tmpJ = Math.min(targetCo, j + problem[3]);

						dp[tmpI][tmpJ] = Math.min(dp[tmpI][tmpJ], dp[i][j] + problem[4]);
					}
				}
			}
		}

        return dp[targetAl][targetCo];
    }
}